File: Bank_Loan_(Erlang_and_Java)/Customer.java


/*
 * @author : Himanshu Kohli
 * @version : 1.0
 * @date 23 June 2019
 * Customer class 
 */

import java.util.HashMap;
import java.util.Random;
import java.util.Set;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;

/*
 * Customer class 
 */
public class Customer implements Runnable {

	DataHolder holder;
	HashMap<Thread,Customer> processList;
	BlockingQueue<String> blockingQueue;
	HashMap<Thread,Banks> banksprocessMap;
	long start_timer;
	
	// Constructor for customers
	public Customer(DataHolder holder, HashMap<Thread,Customer> process,HashMap<Thread,Banks> banksprocessMap) {
		this.holder = holder;
		this.processList = process;	
		blockingQueue = new ArrayBlockingQueue<String>(10000);
		this.banksprocessMap = banksprocessMap;
		
	}
	
	// sends request to the bank for a loan amount
	public synchronized void sendRequest() throws InterruptedException {
		if(!holder.bankNames.isEmpty() && holder.balance > 0) {
			String randBankName = holder.bankNames.get(new Random().nextInt(holder.bankNames.size()));
			Thread.sleep(new Random().nextInt(100));
			
			Set<Thread> keySet = banksprocessMap.keySet();
			for(Thread t : keySet) {
				Banks temp = banksprocessMap.get(t);
				if(t.getName().equalsIgnoreCase(randBankName.trim())) {
					int loan = (new Random().nextInt(holder.balance+1));
						if(loan > 0) {
							temp.bankBlockingQueue.put( holder.name  + "::" + loan);
							Money.MoneyQueue.put(holder.name + " requests a loan of "+ loan + " from " + randBankName);
							
						}
					}
				}
			}
		}

	// listens to the reply from bank class 
	public void listenRequests() throws InterruptedException {
		this.start_timer = System.currentTimeMillis();
		while(true) {
			
				if(!this.blockingQueue.isEmpty() && this.blockingQueue.peek().contains("ACCEPT")) {
					String message = this.blockingQueue.poll();
					String messageBroken[] = message.split(" ");
					int approvedLoan = Integer.parseInt(messageBroken[2].trim());
					this.holder.balance = this.holder.balance - approvedLoan;
					this.start_timer = System.currentTimeMillis();
					sendRequest();
				}
				else if(!this.blockingQueue.isEmpty()  && this.blockingQueue.peek().contains("REJECTS")) {
					String message = this.blockingQueue.poll();
					String messageBroken[] = message.split(" ");
					Money.MoneyQueue.put("[Removing bank "+messageBroken[1].trim() + " from " + holder.name + "]");
					this.holder.bankNames.remove(messageBroken[1].trim());
					this.start_timer = System.currentTimeMillis();
					sendRequest();
				}
				
				if(this.blockingQueue.isEmpty() && (System.currentTimeMillis() - this.start_timer) >= 1500) {
					if(this.holder.balance == 0)
						System.out.println("\nWhooo Hoooo!!!, " + Thread.currentThread().getName() + " has achieved a balance of "+ holder.orignal_balance);
					else
						System.out.println("\nBoooo Hoooo!!!, " + Thread.currentThread().getName() + " has left "+ holder.balance + " to reach a goal of "+ holder.orignal_balance);
					break;
				}		
		}
	}

	
	// overriden run method to execute the threads
	@Override
	public void run() {
		// TODO Auto-generated method stub
		
		try {
			sendRequest();
		} catch (InterruptedException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		
		try {
			listenRequests();
		} catch (InterruptedException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		
	}
}

File: Bank_Loan_(Erlang_and_Java)/customer.erl
%% @author Himanshu Kohli
%% @doc @todo Add description to customer2.


-module(customer).


-export([customer_executor/4]).


customer_executor(Customer_Name,Customer_Balance,Orignal_Balance,List_Of_Bank_Pid) -> 
	receive 
		{Money_PID,C_Name} ->	
							Length = length(List_Of_Bank_Pid),
							
								if Length /= 0 -> RandomBankInfo = lists:nth( rand:uniform(Length) , List_Of_Bank_Pid);
								true -> RandomBankInfo = 0
								end,
							
							Check = checkRandomAndBalance(RandomBankInfo,Customer_Balance),
							if Check ->
									    {RandomBank,B_Name} = RandomBankInfo,
										Loan = generateLoanAmt(Customer_Balance),
										Money_PID ! {customer,C_Name,Loan,B_Name,Customer_Balance},
										timer:sleep(rand:uniform(100)),
										RandomBank ! {Money_PID,self(),Customer_Name,Loan};
							true -> Money_PID ! {customer_result,C_Name,Customer_Balance,Orignal_Balance}
							end,
							customer_executor(Customer_Name, Customer_Balance,Orignal_Balance, List_Of_Bank_Pid);
		
		{bank_reply,approves,Money_PID,Loan_Amount} ->
								self() ! {Money_PID,Customer_Name},
								customer_executor(Customer_Name,Customer_Balance - Loan_Amount,Orignal_Balance,List_Of_Bank_Pid);
		
		{bank_reply,rejects,Money_PID,Bank_Info_Tuple} ->
								UpdateBankList = List_Of_Bank_Pid -- [Bank_Info_Tuple],
								Money_PID ! {bank_removal,Customer_Name,Bank_Info_Tuple,Customer_Balance},
								self() ! {Money_PID,Customer_Name},
								customer_executor(Customer_Name,Customer_Balance,Orignal_Balance,UpdateBankList)
		
		after 1500 -> 
					if Customer_Balance == 0 -> io:fwrite("Whoo hooo, ~p has reached the goal ~p \n",[Customer_Name,Orignal_Balance]);
						true -> io:fwrite("Boo Hoo, ~p has ~p amount left to reach the goal of ~p\n",[Customer_Name,Customer_Balance,Orignal_Balance])
						end
			end. 

generateLoanAmt(X) -> 
				if X >= 50 -> rand:uniform(50);
					true -> rand:uniform(X) 
				end.

checkRandomAndBalance(Rand,Balance) ->
		if Rand /= 0 ->
			   if Balance /= 0 ->	true;
			   true -> false end;
		true -> false end.